crates/core/src/wasm_runtime/contract.rs

```diff
@@ -1,12 +1,35 @@
-use std::{thread::JoinHandle, time::Duration};
+use std::sync::OnceLock;
+use std::time::Duration;
 
 use super::{ContractExecError, RuntimeResult};
 use freenet_stdlib::prelude::{
     ContractInterfaceResult, ContractKey, Parameters, RelatedContracts, StateDelta, StateSummary,
     UpdateData, UpdateModification, ValidateResult, WrappedState,
 };
+use tokio::sync::Semaphore;
 use wasmer::{Instance, Store, TypedFunction};
 
+/// Semaphore to limit concurrent WASM operations.
+///
+/// While tokio's blocking thread pool is bounded (default 512), we use a tighter
+/// limit matching CPU count. This prevents overwhelming the system with concurrent
+/// WASM executions and provides backpressure to callers.
+static WASM_SEMAPHORE: OnceLock<Semaphore> = OnceLock::new();
```

Comment:
this is completely redundant. we already have a pool of executors which controls the parallelism